test/propagate_precision.cpp

```diff
@@ -0,0 +1,158 @@
+        auto y        = m1.add_parameter("y", s2);
+        auto convert1 = m1.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::half_type}}), x);
+        auto two  = m1.add_literal(migraphx::literal{{migraphx::shape::half_type}, {2}});
+        auto div  = migraphx::add_common_op(m1, migraphx::make_op("div"), {convert1, two});
+        auto sqrt = m1.add_instruction(migraphx::make_op("sqrt"), div);
+        auto mul  = m1.add_instruction(migraphx::make_op("mul"), sqrt, y);
+        m1.add_return({mul});
+    }
+    run_pass(m1);
+    migraphx::module m2;
+    {
+        auto x        = m2.add_parameter("x", s1);
+        auto y        = m2.add_parameter("y", s2);
+        auto two      = m2.add_literal(migraphx::literal{{migraphx::shape::half_type}, {2}});
+        auto div      = migraphx::add_common_op(m2, migraphx::make_op("div"), {x, two});
+        auto sqrt     = m2.add_instruction(migraphx::make_op("sqrt"), div);
+        auto convert1 = m2.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::half_type}}), sqrt);
+        auto mul = m2.add_instruction(migraphx::make_op("mul"), convert1, y);
+        m2.add_return({mul});
+    }
+    EXPECT(m1.sort() == m2.sort());
+}
+
+TEST_CASE(propagate_conflict)
+{
+    migraphx::shape s1{migraphx::shape::float_type, {2, 3}};
+    migraphx::shape s2{migraphx::shape::double_type, {2, 3}};
+    migraphx::module m1;
+    {
+        auto x        = m1.add_parameter("x", s1);
+        auto y        = m1.add_parameter("y", s2);
+        auto convert1 = m1.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::half_type}}), x);
+        auto two      = m1.add_literal(migraphx::literal{{migraphx::shape::half_type}, {2}});
+        auto div      = migraphx::add_common_op(m1, migraphx::make_op("div"), {convert1, two});
+        auto sqrt     = m1.add_instruction(migraphx::make_op("sqrt"), div);
+        auto convert2 = m1.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::double_type}}), sqrt);
+        auto mul = m1.add_instruction(migraphx::make_op("mul"), convert2, y);
+        m1.add_return({mul});
+    }
+    run_pass(m1);
+    migraphx::module m2;
+    {
+        auto x        = m2.add_parameter("x", s1);
+        auto y        = m2.add_parameter("y", s2);
+        auto convert1 = m2.add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::double_type}}), x);
+        auto two  = m2.add_literal(migraphx::literal{{migraphx::shape::half_type}, {2}});
+        auto div  = migraphx::add_common_op(m2, migraphx::make_op("div"), {convert1, two});
+        auto sqrt = m2.add_instruction(migraphx::make_op("sqrt"), div);
+        auto mul  = m2.add_instruction(migraphx::make_op("mul"), sqrt, y);
+        m2.add_return({mul});
+    }
+    EXPECT(m1.sort() == m2.sort());
```

Comment:
I understand this is done to preserve precision throughout all these divides/square roots but are we not worried about the added overhead here now? we've just converted the fp16 set of ops to double, or is compute not a concern here?

Reply:
But we are converting to double anyways. For elementwise, the compute shouldn't be that much overhead since these are all essentially unary operators.